File: apps/patter-voice-sdk/libraries/typescript/src/integrations/patter-tool.ts
/**
 * PatterTool — wrap a live Patter instance as a tool callable from external
 * agent frameworks (OpenAI Assistants, Anthropic Claude tool-use, LangChain,
 * Hermes Agent, MCP, generic OpenAI-compatible endpoints).
 *
 * Pattern this enables: a customer already runs an agent in their existing
 * stack (LangChain, OpenAI Assistant, Hermes Agent, …) and wants the agent
 * to *make phone calls* during a conversation. With this tool, the customer
 * registers `make_phone_call` and the agent's tool-call loop can dial out
 * via Patter, get a transcript + cost back, and continue reasoning.
 *
 * ## Design
 *
 * Each `PatterTool` wraps one `Patter` instance (carrier + agent + serve).
 * The tool exposes:
 *
 *   - `openaiSchema()`     — OpenAI / chat-completions tool spec
 *   - `anthropicSchema()`  — Anthropic Claude tool spec
 *   - `hermesSchema()`     — Hermes Agent / Nous registry schema (alias for
 *                            anthropicSchema; same JSON-Schema shape)
 *   - `execute(args)`      — dial outbound, await call end, return summary
 *   - `hermesHandler()`    — `(args, **kw) => Promise<string>` wrapper that
 *                            returns a JSON string and `{"error": "..."}` on
 *                            failure (matches Hermes' tool contract)
 *
 * ## Usage (OpenAI / Anthropic)
 *
 * ```ts
 * import { Patter, Twilio, DeepgramSTT, GroqLLM, ElevenLabsTTS } from 'getpatter';
 * import { PatterTool } from 'getpatter/integrations';
 *
 * const phone = new Patter({
 *   carrier: new Twilio(),
 *   phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
 *   webhookUrl: 'agent.example.com',
 * });
 *
 * const tool = new PatterTool({
 *   phone,
 *   agent: { stt: new DeepgramSTT(), llm: new GroqLLM(), tts: new ElevenLabsTTS() },
 * });
 *
 * await tool.start();   // boots phone.serve() once
 *
 * // Register with your LLM
 * const tools = [tool.openaiSchema()];
 *
 * // When the LLM emits a tool_call:
 * const result = await tool.execute({
 *   to: '+15551234567',
 *   goal: 'Book a dentist appointment for next Tuesday afternoon.',
 * });
 * // → { call_id, status, duration_seconds, cost_usd, transcript, … }
 * ```
 *
 * ## Usage (Hermes Agent)
 *
 * Hermes' contract: handler takes `args: dict` + kwargs, returns a JSON
 * string. The TS SDK is meant to be invoked from Python via your own bridge
 * (HTTP, MCP, subprocess); this `hermesSchema()` + `hermesHandler()` pair
 * matches the Python adapter shipped under `getpatter.integrations` so the
 * two SDKs stay in lockstep.
 *
 * For pure-Python Hermes setups, use `PatterTool` from `getpatter.integrations`
 * directly inside a `tools/patter.py` module:
 *
 * ```python
 * from tools.registry import registry
 * from getpatter.integrations import PatterTool
 *
 * tool = PatterTool(phone=...)
 * tool.register_hermes(registry)
 * ```
 */

import type { Patter } from '../client';
import type { AgentOptions, CallResult } from '../types';

/** JSON-Schema of the call args. Identical wire shape across openai/anthropic/hermes. */
const PARAMETERS_SCHEMA = {
  type: 'object' as const,
  properties: {
    to: {
      type: 'string',
      description:
        'Destination phone number in E.164 format (e.g. "+15551234567"). Required.',
    },
    goal: {
      type: 'string',
      description:
        "What the agent should accomplish on the call. Becomes the in-call agent's system prompt for this single call.",
    },
    first_message: {
      type: 'string',
      description:
        'Optional first message the agent speaks when the callee answers. Defaults to a generic greeting.',
    },
    max_duration_sec: {
      type: 'integer',
      description:
        'Hard timeout for the call in seconds. Default 180. The call is force-ended at this deadline whether or not it has resolved.',
      minimum: 5,
      maximum: 1800,
    },
  },
  required: ['to'],
} as const;

const DEFAULT_NAME = 'make_phone_call';
const DEFAULT_DESCRIPTION =
  'Place a real outbound phone call. Returns a JSON object with the full transcript, ' +
  'call status, duration in seconds, and cost. Use this when the user asks you to call ' +
  'someone, schedule appointments by phone, or otherwise reach a human via voice.';

/** Constructor options for `PatterTool`. */
export interface PatterToolOptions {
  /**
   * Patter instance to dial through. Must be in local mode (have a `carrier`).
   * The tool boots `phone.serve()` on `start()`; do not call `serve()` yourself.
   */
  readonly phone: Patter;
  /**
   * Default agent config used for outbound calls. Per-call overrides come from
   * `execute({ goal, first_message })`.
   */
  readonly agent?: AgentOptions;
  /** Tool name shown to the LLM. Default `'make_phone_call'`. */
  readonly name?: string;
  /** Tool description for the LLM. Default tuned for English assistants. */
  readonly description?: string;
  /** Default per-call timeout in seconds. Default 180. */
  readonly maxDurationSec?: number;
  /**
   * Optional pass-through for `phone.serve()`'s `recording` flag — record all
   * outbound calls placed via this tool.
   */
  readonly recording?: boolean;
}

/** Args accepted by `PatterTool.execute()` (and the OpenAI/Anthropic/Hermes tool schemas). */
export interface PatterToolExecuteArgs {
  readonly to: string;
  readonly goal?: string;
  readonly first_message?: string;
  readonly max_duration_sec?: number;
}

/** Result envelope returned by `PatterTool.execute()` once the underlying call ends. */
export interface PatterToolResult {
  readonly call_id: string;
  readonly status: string;
  readonly duration_seconds: number;
  /**
   * Carrier-agnostic outcome (answered / voicemail / no_answer / busy /
   * failed) lifted from the SDK {@link CallResult}. Optional for backward
   * compatibility with any code constructing this envelope without it.
   */
  readonly outcome?: string;
  readonly cost_usd?: number;
  readonly transcript: ReadonlyArray<
    Readonly<{ role: string; text: string; timestamp?: number }>
  >;
  readonly metrics?: Readonly<Record<string, unknown>> | null;
}

/** Wraps a live `Patter` instance as a tool callable from external agent frameworks. */
export class PatterTool {
  readonly name: string;
  readonly description: string;
  private readonly phone: Patter;
  private readonly agent: AgentOptions | undefined;
  private readonly maxDurationSec: number;
  private readonly recording: boolean;
  private started = false;
  private hermesTelemetryEmitted = false;
  /** Cached in-progress (or completed) start promise so concurrent execute()
   *  callers all await the same boot sequence instead of each racing into
   *  phone.serve(). Reset to null on failure so callers can retry after a
   *  transient error. */
  private startPromise: Promise<void> | null = null;

  constructor(opts: PatterToolOptions) {
    if (!opts.phone) {
      throw new Error('PatterTool: `phone` (a Patter instance) is required.');
    }
    this.phone = opts.phone;
    this.agent = opts.agent;
    this.name = opts.name ?? DEFAULT_NAME;
    this.description = opts.description ?? DEFAULT_DESCRIPTION;
    this.maxDurationSec = Math.max(5, Math.min(1800, opts.maxDurationSec ?? 180));
    this.recording = opts.recording ?? false;
  }

  // --- Schema exporters ---------------------------------------------------

  /** OpenAI Chat Completions / Assistants tool spec. */
  openaiSchema(): {
    type: 'function';
    function: { name: string; description: string; parameters: typeof PARAMETERS_SCHEMA };
  } {
    return {
      type: 'function',
      function: {
        name: this.name,
        description: this.description,
        parameters: PARAMETERS_SCHEMA,
      },
    };
  }

  /** Anthropic Messages API tool spec. */
  anthropicSchema(): {
    name: string;
    description: string;
    input_schema: typeof PARAMETERS_SCHEMA;
  } {
    return {
      name: this.name,
      description: this.description,
      input_schema: PARAMETERS_SCHEMA,
    };
  }

  /**
   * Hermes Agent (Nous Research) registry schema. Same JSON-Schema shape as
   * Anthropic's; Hermes consumes it via `registry.register({ schema: ... })`.
   */
  hermesSchema(): {
    name: string;
    description: string;
    parameters: typeof PARAMETERS_SCHEMA;
  } {
    return {
      name: this.name,
      description: this.description,
      parameters: PARAMETERS_SCHEMA,
    };
  }

  // --- Lifecycle ----------------------------------------------------------

  /**
   * Start the underlying Patter server. Idempotent.
   *
   * `execute()` relies on `Patter.call({ wait: true })`, which requires an
   * active server to receive the carrier completion webhooks — that's what
   * `serve()` provides here. No `onCallEnd` callback is wired: the SDK's own
   * per-callId completion registry resolves the result, so the user's
   * `onCallEnd` slot is left free.
   *
   * Idempotent and concurrency-safe: concurrent callers all await the same
   * in-progress boot instead of each racing into `phone.serve()`.
   */
  async start(): Promise<void> {
    if (this.startPromise) return this.startPromise;
    this.startPromise = this._doStart();
    try {
      await this.startPromise;
    } catch (err) {
      // Allow retry after a transient failure.
      this.startPromise = null;
      throw err;
    }
  }

  private async _doStart(): Promise<void> {
    if (this.started) return;
    if (!this.agent) {
      throw new Error(
        'PatterTool.start: `agent` config is required. Pass `{ stt, llm, tts }` ' +
          'or an `engine` (e.g. OpenAIRealtime) when constructing PatterTool.',
      );
    }
    const builtAgent = this.phone.agent(this.agent);
    await this.phone.serve({
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      agent: builtAgent as any,
      recording: this.recording,
    });
    this.started = true;
  }

  /** Best-effort shutdown — tear the Patter server down via `disconnect()`. */
  async stop(): Promise<void> {
    if (!this.started) return;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const disconnectable = this.phone as unknown as { disconnect?: () => Promise<void> };
    if (typeof disconnectable.disconnect === 'function') {
      try {
        await disconnectable.disconnect();
      } catch {
        /* defensive — shutdown must not throw */
      }
    }
    this.started = false;
    this.startPromise = null;
  }

  // --- Execution ----------------------------------------------------------

  /**
   * Dial outbound, wait for the call to end, return a structured result.
   *
   * Thin wrapper over `Patter.call({ wait: true })`: the SDK now owns the
   * dial → callId → terminal-signal correlation, so this just bounds the wait
   * with `max_duration_sec` and maps the {@link CallResult} into the tool's
   * public envelope. Mirrors Python's `PatterTool.execute`.
   */
  async execute(args: PatterToolExecuteArgs): Promise<PatterToolResult> {
    if (!this.started) await this.start();
    if (!args || typeof args.to !== 'string' || !args.to.startsWith('+')) {
      throw new Error('PatterTool.execute: `to` must be an E.164 phone number (e.g. "+15551234567").');
    }
    const timeoutSec = Math.max(
      5,
      Math.min(1800, args.max_duration_sec ?? this.maxDurationSec),
    );

    const baseAgent = this.agent ?? ({} as AgentOptions);
    const overrideAgent = this.phone.agent({
      ...baseAgent,
      ...(args.goal !== undefined ? { systemPrompt: args.goal } : {}),
      ...(args.first_message !== undefined ? { firstMessage: args.first_message } : {}),
    });

    let timer: NodeJS.Timeout | undefined;
    const timeout = new Promise<never>((_resolve, reject) => {
      timer = setTimeout(() => {
        reject(
          new Error(
            `PatterTool.execute: call to ${args.to} exceeded ${timeoutSec}s timeout`,
          ),
        );
      }, timeoutSec * 1000);
      timer.unref?.();
    });

    let result: CallResult | void;
    try {
      result = await Promise.race([
        this.phone.call({
          to: args.to,
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          agent: overrideAgent as any,
          wait: true,
        }),
        timeout,
      ]);
    } finally {
      if (timer) clearTimeout(timer);
    }

    return resultFromCallResult(result);
  }

  /**
   * Hermes-style handler: `(args, kwargs) => Promise<string>` returning a JSON
   * string with either the result envelope or an `{"error": "..."}` payload.
   * Mirrors the Python `PatterTool.hermes_handler` so cross-SDK adapters share
   * the same wire contract.
   */
  hermesHandler(): (args: PatterToolExecuteArgs) => Promise<string> {
    // Anonymous telemetry: Patter is being exposed as a tool to a Hermes agent.
    if (!this.hermesTelemetryEmitted) {
      this.hermesTelemetryEmitted = true;
      try {
        const tel = (
          this.phone as unknown as {
            telemetry?: { record(name: string, dims?: Record<string, unknown>): void };
          }
        ).telemetry;
        // Emit on agent_configured (where `integration` is a native dimension)
        // with the full shape so it does not create a second feature_used shape.
        tel?.record('agent_configured', {
          builtin_tool_count: 0,
          custom_tool_count_bucket: '0',
          integration: 'hermes',
          integration_kind: 'none',
          mcp_server_count_bucket: '0',
        });
      } catch {
        /* ignore — telemetry never breaks integration */
      }
    }
    return async (args: PatterToolExecuteArgs) => {
      try {
        const result = await this.execute(args);
        return JSON.stringify(result);
      } catch (err) {
        return JSON.stringify({ error: err instanceof Error ? err.message : String(err) });
      }
    };
  }
}

/**
 * Map an SDK {@link CallResult} into the tool's public envelope.
 *
 * Reads structured fields off the result directly — `cost.total` and
 * `durationSeconds` are real numbers here. `metrics` is passed through as a
 * plain object so JSON serialization for the Hermes/MCP wire envelope stays
 * clean. Mirrors Python's `_result_from_call_result`.
 */
function resultFromCallResult(result: CallResult | void): PatterToolResult {
  if (!result) {
    // call({ wait: true }) always resolves to a CallResult; this guard only
    // exists because the union type includes void. Treat the impossible case
    // as an empty completed result rather than throwing.
    return {
      call_id: '',
      status: 'completed',
      outcome: '',
      duration_seconds: 0,
      cost_usd: undefined,
      transcript: [],
      metrics: null,
    };
  }
  const costTotal = result.cost?.total;
  const costUsd = typeof costTotal === 'number' ? costTotal : undefined;
  const metrics = result.metrics
    ? (result.metrics as unknown as Record<string, unknown>)
    : null;
  return {
    call_id: result.callId || '',
    status: result.status || 'completed',
    outcome: result.outcome || '',
    duration_seconds:
      typeof result.durationSeconds === 'number' ? result.durationSeconds : 0,
    cost_usd: costUsd,
    transcript: result.transcript ? [...result.transcript] : [],
    metrics,
  };
}

File: apps/patter-voice-sdk/libraries/python/getpatter/integrations/patter_tool.py
"""PatterTool — wrap a live Patter instance as a tool callable from external
agent frameworks (OpenAI Assistants, Anthropic Claude tool-use, LangChain,
Hermes Agent, MCP, generic OpenAI-compatible endpoints).

See ``libraries/typescript/src/integrations/patter-tool.ts`` for the matching TS module —
the wire contracts (parameter schema, result envelope, ``hermes_handler``
return shape) are kept identical so a customer can swap SDKs at any time.

Usage (Hermes Agent integration)::

    # in your tools/patter.py file (auto-discovered by Hermes)
    from tools.registry import registry
    from getpatter import Patter, Twilio, DeepgramSTT, GroqLLM, ElevenLabsTTS
    from getpatter.integrations import PatterTool

    phone = Patter(
        carrier=Twilio(),
        phone_number=os.environ["TWILIO_PHONE_NUMBER"],
        webhook_url="agent.example.com",
    )

    tool = PatterTool(
        phone=phone,
        agent={"stt": DeepgramSTT(), "llm": GroqLLM(), "tts": ElevenLabsTTS()},
    )

    # One-line registration with Hermes' registry:
    tool.register_hermes(registry)

Usage (OpenAI / Anthropic / generic)::

    tool = PatterTool(phone=phone, agent={...})
    schemas = {"openai": tool.openai_schema(), "anthropic": tool.anthropic_schema()}

    # When the LLM emits a tool_call:
    result = await tool.execute(to="+15551234567", goal="Book a dentist appointment")
    # → {"call_id": "...", "status": "completed", "duration_seconds": 12.3, ...}
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import asdict, dataclass, field, is_dataclass
from typing import Any, Awaitable, Callable

logger = logging.getLogger("getpatter.integrations.patter_tool")


# JSON-Schema for the call args. Identical wire shape across openai/anthropic/hermes.
_PARAMETERS_SCHEMA: dict[str, Any] = {
    "type": "object",
    "properties": {
        "to": {
            "type": "string",
            "description": (
                'Destination phone number in E.164 format (e.g. "+15551234567"). Required.'
            ),
        },
        "goal": {
            "type": "string",
            "description": (
                "What the agent should accomplish on the call. Becomes the in-call "
                "agent's system prompt for this single call."
            ),
        },
        "first_message": {
            "type": "string",
            "description": (
                "Optional first message the agent speaks when the callee answers. "
                "Defaults to a generic greeting."
            ),
        },
        "max_duration_sec": {
            "type": "integer",
            "description": (
                "Hard timeout for the call in seconds. Default 180. The call is "
                "force-ended at this deadline whether or not it has resolved."
            ),
            "minimum": 5,
            "maximum": 1800,
        },
    },
    "required": ["to"],
}

_DEFAULT_NAME = "make_phone_call"
_DEFAULT_DESCRIPTION = (
    "Place a real outbound phone call. Returns a JSON object with the full "
    "transcript, call status, duration in seconds, and cost. Use this when the "
    "user asks you to call someone, schedule appointments by phone, or otherwise "
    "reach a human via voice."
)


@dataclass(frozen=True)
class PatterToolResult:
    """Structured result returned by ``PatterTool.execute``."""

    call_id: str
    status: str
    duration_seconds: float
    # Carrier-agnostic outcome (answered / voicemail / no_answer / busy /
    # failed) lifted from the SDK ``CallResult``. Defaulted for backward
    # compatibility with any code constructing this envelope positionally.
    outcome: str = ""
    transcript: tuple[dict[str, Any], ...] = field(default_factory=tuple)
    cost_usd: float | None = None
    metrics: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class PatterTool:
    """Wrap a live ``Patter`` instance as a tool callable from external agent
    frameworks. Schema-stable across OpenAI / Anthropic / Hermes; ``execute()``
    runs an outbound dial and waits for completion.
    """

    def __init__(
        self,
        phone: Any,
        agent: dict[str, Any] | None = None,
        name: str = _DEFAULT_NAME,
        description: str = _DEFAULT_DESCRIPTION,
        max_duration_sec: int = 180,
        recording: bool = False,
    ) -> None:
        if phone is None:
            raise ValueError("PatterTool: `phone` (a Patter instance) is required.")
        self._phone = phone
        self._agent_spec = agent
        self.name = name
        self.description = description
        self._max_duration_sec = max(5, min(1800, int(max_duration_sec)))
        self._recording = recording
        self._started = False

    # --- Schema exporters --------------------------------------------------

    def openai_schema(self) -> dict[str, Any]:
        """OpenAI Chat Completions / Assistants tool spec."""
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": _PARAMETERS_SCHEMA,
            },
        }

    def anthropic_schema(self) -> dict[str, Any]:
        """Anthropic Messages API tool spec."""
        return {
            "name": self.name,
            "description": self.description,
            "input_schema": _PARAMETERS_SCHEMA,
        }

    def hermes_schema(self) -> dict[str, Any]:
        """Hermes Agent (Nous Research) registry schema. Same JSON-Schema as
        Anthropic's, exposed under ``parameters`` to match Hermes' contract."""
        return {
            "name": self.name,
            "description": self.description,
            "parameters": _PARAMETERS_SCHEMA,
        }

    # --- Hermes registry helper -------------------------------------------

    def register_hermes(self, registry: Any, toolset: str = "patter") -> None:
        """Register this tool with a Hermes ``tools.registry.Registry`` instance.

        Hermes' tool contract is ``handler(args: dict, **kw) -> str`` returning
        a JSON string (errors as ``{"error": "..."}``). We bridge it to
        ``execute()``.
        """
        if not hasattr(registry, "register"):
            raise TypeError(
                "PatterTool.register_hermes: argument must be a Hermes Registry "
                "(needs a `.register()` method)."
            )
        handler = self.hermes_handler()
        registry.register(
            name=self.name,
            toolset=toolset,
            schema=self.hermes_schema(),
            handler=handler,
        )
        # Anonymous telemetry: Patter is being exposed as a tool to a Hermes agent.
        # Emit on ``agent_configured`` (where ``integration`` is a native dimension,
        # alongside openclaw/mcp) with the full shape so it does not create a second
        # ``feature_used`` shape.
        try:
            tel = getattr(self._phone, "_telemetry", None)
            if tel is not None:
                tel.record(
                    "agent_configured",
                    builtin_tool_count=0,
                    custom_tool_count_bucket="0",
                    integration="hermes",
                    integration_kind="none",
                    mcp_server_count_bucket="0",
                )
        except Exception:
            pass

    # --- Lifecycle ---------------------------------------------------------

    async def start(self) -> None:
        """Start the underlying Patter server. Idempotent.

        ``execute()`` relies on ``Patter.call(wait=True)``, which requires an
        active server to receive the carrier completion webhooks — that's what
        ``serve()`` provides here. No ``on_call_end`` callback is wired: the
        SDK's own per-call_id completion registry resolves the result, so the
        user's ``on_call_end`` slot is left free.
        """
        if self._started:
            return
        if not self._agent_spec:
            raise ValueError(
                "PatterTool.start: `agent` config is required. Pass "
                "`{'stt': ..., 'llm': ..., 'tts': ...}` or an `engine` "
                "(e.g. OpenAIRealtime) when constructing PatterTool."
            )
        built_agent = self._phone.agent(**self._agent_spec)
        await self._phone.serve(agent=built_agent, recording=self._recording)
        self._started = True

    async def stop(self) -> None:
        """Best-effort shutdown — tear the Patter server down via disconnect()."""
        if not self._started:
            return
        if hasattr(self._phone, "disconnect"):
            try:
                await self._phone.disconnect()
            except Exception:  # pragma: no cover - defensive
                logger.debug(
                    "PatterTool.stop: phone.disconnect() failed", exc_info=True
                )
        self._started = False

    # --- Execution ---------------------------------------------------------

    async def execute(
        self,
        to: str,
        goal: str | None = None,
        first_message: str | None = None,
        max_duration_sec: int | None = None,
    ) -> PatterToolResult:
        """Dial outbound, wait for the call to end, return a structured result.

        Thin wrapper over ``Patter.call(wait=True)``: the SDK now owns the
        dial → call_id → terminal-signal correlation, so this just bounds the
        wait with ``max_duration_sec`` and maps the ``CallResult`` into the
        tool's public envelope.
        """
        if not isinstance(to, str) or not to.startswith("+"):
            raise ValueError(
                'PatterTool.execute: `to` must be an E.164 phone number (e.g. "+15551234567").'
            )
        if not self._started:
            await self.start()
        timeout = max(5, min(1800, int(max_duration_sec or self._max_duration_sec)))

        agent_kwargs = dict(self._agent_spec or {})
        if goal is not None:
            agent_kwargs["system_prompt"] = goal
        if first_message is not None:
            agent_kwargs["first_message"] = first_message
        override_agent = self._phone.agent(**agent_kwargs)

        try:
            result = await asyncio.wait_for(
                self._phone.call(to=to, agent=override_agent, wait=True),
                timeout=timeout,
            )
        except asyncio.TimeoutError as exc:
            raise TimeoutError(
                f"PatterTool.execute: call to {to} exceeded {timeout}s timeout"
            ) from exc

        return _result_from_call_result(result)

    def hermes_handler(self) -> Callable[..., Awaitable[str]]:
        """Return a Hermes-compatible handler ``(args, **kw) -> Awaitable[str]``.

        The handler returns a JSON string with the result envelope, or
        ``{"error": "..."}`` if execution failed. Matches Hermes' contract
        (see ``hermes-agent.nousresearch.com/docs/developer-guide/adding-tools``).
        """

        async def handler(args: dict[str, Any], **_kw: Any) -> str:
            try:
                result = await self.execute(
                    to=args.get("to") or "",
                    goal=args.get("goal"),
                    first_message=args.get("first_message"),
                    max_duration_sec=args.get("max_duration_sec"),
                )
                return json.dumps(result.to_dict(), default=str)
            except Exception as exc:
                return json.dumps({"error": str(exc)})

        return handler


def _result_from_call_result(result: Any) -> PatterToolResult:
    """Map an SDK ``CallResult`` into the tool's public envelope.

    Reads structured attributes off the dataclass directly — ``cost.total``
    and ``duration_seconds`` are real numbers here (the previous dict-probing
    implementation silently dropped both because the live payload delivers a
    ``CallMetrics`` dataclass, not a dict). ``metrics`` is flattened to a plain
    dict so ``to_dict()`` stays JSON-friendly for the Hermes/MCP wire envelope.
    """
    cost_obj = getattr(result, "cost", None)
    cost_total = getattr(cost_obj, "total", None)
    cost_usd = float(cost_total) if isinstance(cost_total, (int, float)) else None

    metrics_obj = getattr(result, "metrics", None)
    if is_dataclass(metrics_obj):
        metrics_dict: dict[str, Any] | None = asdict(metrics_obj)
    elif isinstance(metrics_obj, dict):
        metrics_dict = metrics_obj
    else:
        metrics_dict = None

    transcript = tuple(getattr(result, "transcript", ()) or ())
    return PatterToolResult(
        call_id=getattr(result, "call_id", ""),
        status=str(getattr(result, "status", "") or "completed"),
        outcome=str(getattr(result, "outcome", "") or ""),
        duration_seconds=float(getattr(result, "duration_seconds", 0.0) or 0.0),
        cost_usd=cost_usd,
        transcript=transcript,
        metrics=metrics_dict,
    )
